runtime/src/current_wasm_module_instance.c

```diff
@@ -27,5 +27,5 @@ sledge_abi__current_wasm_module_instance_trap(enum sledge_abi__wasm_trap trapno)
 	assert(sandbox != NULL);
 	assert(sandbox->state == SANDBOX_RUNNING_USER || sandbox->state == SANDBOX_RUNNING_SYS);
 
-	longjmp(sandbox->ctxt.start_buf, trapno);
+	siglongjmp(sandbox->ctxt.start_buf, trapno);
```

Comment:
I'm not sure I can build the context to understand all of this right now, and I think that @emil916 should understand the control flow/intention here. 

Looks good to me, but I suggest that he eventually validate.

Reply:
I'd imagine that you're reporting/recording the failure at the `setjmp` site.

Reply:
Pretty straight-forward.

```c
void
current_sandbox_start(void)
{
	struct sandbox *sandbox = current_sandbox_init();

	int rc = setjmp(sandbox->ctxt.start_buf);
	if (rc == 0) {
		struct module *current_module = sandbox_get_module(sandbox);
		sandbox->return_value         = module_entrypoint(current_module);
	} else {
		current_sandbox_wasm_trap_handler(rc);
	}

	current_sandbox_fini();
}
```

```c
void
current_sandbox_wasm_trap_handler(int trapno)
{
	char           *error_message = NULL;
	struct sandbox *sandbox       = current_sandbox_get();
	sandbox_syscall(sandbox);

	switch (trapno) {
	case WASM_TRAP_EXIT:
		break;
	case WASM_TRAP_INVALID_INDEX:
		error_message = "WebAssembly Trap: Invalid Index\n";
		client_socket_send(sandbox->client_socket_descriptor, http_header_build(500), http_header_len(500),
		                   current_sandbox_sleep);
		break;
	case WASM_TRAP_MISMATCHED_TYPE:
		error_message = "WebAssembly Trap: Mismatched Type\n";
		client_socket_send(sandbox->client_socket_descriptor, http_header_build(500), http_header_len(500),
		                   current_sandbox_sleep);
		break;
	case WASM_TRAP_PROTECTED_CALL_STACK_OVERFLOW:
		error_message = "WebAssembly Trap: Protected Call Stack Overflow\n";
		client_socket_send(sandbox->client_socket_descriptor, http_header_build(500), http_header_len(500),
		                   current_sandbox_sleep);
		break;
	case WASM_TRAP_OUT_OF_BOUNDS_LINEAR_MEMORY:
		error_message = "WebAssembly Trap: Out of Bounds Linear Memory Access\n";
		client_socket_send(sandbox->client_socket_descriptor, http_header_build(500), http_header_len(500),
		                   current_sandbox_sleep);
		break;
	case WASM_TRAP_ILLEGAL_ARITHMETIC_OPERATION:
		error_message = "WebAssembly Trap: Illegal Arithmetic Operation\n";
		client_socket_send(sandbox->client_socket_descriptor, http_header_build(500), http_header_len(500),
		                   current_sandbox_sleep);
		break;
	}

	debuglog("%s", error_message);
	sandbox_close_http(sandbox);
	generic_thread_dump_lock_overhead();
	current_sandbox_exit();
	assert(0);
}
```

```c
void
current_sandbox_exit()
{
	struct sandbox *exiting_sandbox = current_sandbox_get();
	assert(exiting_sandbox != NULL);

	generic_thread_dump_lock_overhead();

	switch (exiting_sandbox->state) {
	case SANDBOX_RETURNED:
		sandbox_exit_success(exiting_sandbox);
		break;
	case SANDBOX_RUNNING_SYS:
		sandbox_exit_error(exiting_sandbox);
		break;
	default:
		panic("Cooperatively switching from a sandbox in a non-terminal %s state\n",
		      sandbox_state_stringify(exiting_sandbox->state));
	}

	scheduler_cooperative_sched(true);

	/* The scheduler should never switch back to completed sandboxes */
	assert(0);
}
```

Reply:
Sorry, stale comment. I ended up chasing this path down. Look fantastic.